lib/omniauth/strategies/saml.rb

```diff
@@ -67,21 +63,67 @@ def callback_phase
             new_hash[key.to_sym] = value
             new_hash
           end
-        response = OneLogin::RubySaml::Response.new(request.params['SAMLResponse'], opts.merge(settings: settings))
-        response.attributes['fingerprint'] = options.idp_cert_fingerprint
 
-        # will raise an error since we are not in soft mode
+        if request.params["SAMLRequest"]
+          logout_request = OneLogin::RubySaml::SloLogoutrequest.new(request.params["SAMLRequest"])
+
+          if logout_request.document.root.name == "LogoutRequest"
+            if logout_request.is_valid?
+              if logout_request.name_id == session["saml_uid"]
+                # Actually log out this session
+                session.clear
+
+                # Generate a response to the IdP.
+                logout_request_id = logout_request.id
+                logout_response = OneLogin::RubySaml::SloLogoutresponse.new.create(settings, logout_request_id, nil, RelayState: request.params["RelayState"])
+                return redirect(logout_response)
+              end
+            end
+          end
+          raise OmniAuth::Strategies::SAML::ValidationError.new("IdP initiated LogoutRequest was not valid!")
+        end
+
+        unless request.params['SAMLResponse']
```

Comment:
Instead of using an unless here and checking for `request.params["SAMLRequest"]` again, could you use an else clause from the original `if` and raise the exception there?